[#44946, #44947] Implementing `Query#timestamps`

fiedl opened the pull request:
In the context of the [baseline-comparison feature](https://community.openproject.org/projects/openproject/work_packages/26448), this pull request enables the `Query` model to query work packages at different points in time.

### Context

![Component_Overview-4](https://user-images.githubusercontent.com/1679688/217393664-8009bca8-39e1-447e-9f17-49a72e54d721.png)

### Timestamps

For the baseline-comparison feature we need to query historic data of work packages. In addition to the current state of the work packages (today), we can ask for the state of work packages at other points in time (**timestamps**).

The timestamps can be specified

- **absolute**, e.g. "2022-08-01", or 
- **relative**, e.g. "1 month ago".

This pull request implements an abstraction class `Timestamp` that can be initialised with [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601) strings.

```ruby
absolute_timestamp = Timestamp.parse("2022-10-29T23:01:23Z")
relative_timestamp = Timestamp.parse("P-5Y")  # 5 years ago
```

The timestamps can be converted to regular times, which are always absolute then:

```ruby
relative_timestamp.to_time
```

The timestamps can be passed to work package queries:

```ruby
WorkPackage.at_timestamp(relative_timestamp).where(...)
```

### Querying without timestamps

This is how one would search work packages with a description containing the expression "foo", which also works without this pull request:

```ruby
query = Query.new
query.add_filter 'description', '~', "foo"
query.results.work_packages
```

This searches for work packages that have the term "foo" in their description in their most current states, i.e. **today**.

### Adding timestamps to a Query

For the baseline-comparison feature, we would like to also search historic data, i.e. find those work packages that

- have the term "foo" in their description today **or**
- had the term "foo" in their description at an earlier point in time (baseline date).

```ruby
today = Time.zone.now
baseline_date = 1.year.ago

query = Query.new
query.add_filter 'description', '~', "foo"
query.timestamps = [baseline_date, today]
query.results.work_packages
```

This returns the work packages in their today's state.

In order to get the historic state of a specific work package at the `baseline_date`, call:

```ruby
work_package = query.results.work_packages.first
work_package.at_timestamp(baseline_date)
```

In order to get the historic state of all the work packages in the query result at the `baseline_date`, call:

```ruby
work_packages = query.results.work_packages
work_packages.at_timestamp(baseline_date)
```

### Query only for historic data

If you are not interested in the work pacakges as they are today, but only in their historic state, configure the query like this:

```ruby
historic_date = 1.year.ago

query = Query.new
query.add_filter 'description', '~', "foo"
query.timestamps = [historic_date]
query.results.work_packages
```

This still returns the the work packages in their current state. To get the work packages in their historic state, call:

```
work_packages = query.results.work_packages
work_packages.at_timestamp(historic_date)
```

Also, please note that while the filters use the given timestamps, sorting is done based only on the current state regardless of the given timestamps.

### Work to be done

- [x] Implement `Timestamp` model as abstraction for absolute and relative timestamps ➝ https://github.com/fiedl/openproject/pull/2/commits/49d14e6631a43fcb15e9bd823db894e8004eea21
- [x] Add serialized `timestamps` attribute to the `Query` model ➝ https://github.com/fiedl/openproject/pull/2/commits/49d14e6631a43fcb15e9bd823db894e8004eea21
- [x] Query results: Search work packages at all `timestamps` ➝ https://github.com/fiedl/openproject/pull/2/commits/2f51f91f23f78d0130bd952e2e24d5f53f07d89e
- [x] Make sure that the previous model specs are still green
- [x] [Check other unit tests](https://github.com/opf/openproject/pull/11678#issuecomment-1326171087)
- [x] [Check feature specs](https://github.com/opf/openproject/pull/11678#issuecomment-1328011996)
- [x] Adding specs with multiple work packages ➝ https://github.com/opf/openproject/pull/11678/commits/f3e41a7a5c17a09eaef406a1394219cc9cb30ab8
- [x] Handle case where the last timestamp is not "now"
- [x] Test, document, and fix [querying the historic state](https://github.com/opf/openproject/pull/11678#issuecomment-1324244907) ➝ https://github.com/fiedl/openproject/commit/47c45b6a9f8a210eaa59fa7139f31cb69eb016fd
- [x] Use sub queries rather than storing the ids ➝ https://github.com/opf/openproject/pull/11678/commits/4ba3c5aa24e64cae7b473c5bf67b36a0a36a3cb4
- [x] After we are sure that the code works, i.e. the specs are green, implement switch to circumvent the `at_timestamp` code when performing a non-historic query ➝ https://github.com/opf/openproject/pull/11678/commits/f551359a14f4b2788ce3ee4ee30913b7d05f7c79

### See also

- **Implementation Overview**: https://github.com/fiedl/openproject/issues/4
- **Epic**: https://community.openproject.org/projects/openproject/work_packages/26448
- Implementation tasks:
  - https://community.openproject.org/projects/openproject/work_packages/44946
  - https://community.openproject.org/projects/openproject/work_packages/44947

fiedl committed:
adding and persisting attribute `Query#timestamps`

in preparation of the baseline-comparison feature

https://community.openproject.org/projects/openproject/work_packages/26448

fiedl committed:
`Query::Results`: search work packages at all given timestamps

The result includes work packages that match the filter criteria at *any* of the given timestamps. In the context of the baseline-comparison feature, this means that the result contains the work packages that do match the filter criteria today, as well as the work packages that have matched the filter criteria at the baseline date.

fiedl committed:
Merge branch https://github.com/opf/openproject/pull/11243 into https://github.com/fiedl/openproject/pull/2 in order to get

- https://github.com/opf/openproject/pull/11243/commits/56984157f511da6b72f5e7f6c5445cde7292d989
- https://github.com/opf/openproject/pull/11243/commits/6f3bd4706fb0398351488419fcd7010763f59f35
- https://github.com/opf/openproject/pull/11243/commits/ebad1063fe2c95e806ebceab39f65189184ec25e

fiedl committed:
Query::Results: include joins and scope in `work_packages`

because otherwise, the sources are missing in the resulting sql statement.

fiedl committed:
Merge https://github.com/opf/openproject/pull/11243 into https://github.com/fiedl/openproject/pull/2

in order to get sub-query support
https://github.com/fiedl/openproject/issues/3

fiedl committed:
Merge https://github.com/opf/openproject/pull/11243 into https://github.com/opf/openproject/pull/11678

fiedl commented:
## Should `Query::Results` return work packages in their historic state?

```ruby
historic_date = 1.year.ago

query = Query.new
query.add_filter 'description', '~', "foo"
query.timestamps = [historic_date]
query.results.work_packages
```

Would we expect `query.results.work_packages` to return the matching work packages in their state at the `historic_date` or just the current work packages?

### (a) Return historic state

My first guess was that they should be in their historic states.

```ruby
class Query::Results
  def work_packages
    if query.historic?
      work_packages_in_historic_state(timestamp: query.timestamps.last)
    else
      work_packages_in_current_state
    end
  end

  def work_packages_in_current_state
    work_package_scope
      .where(id: work_package_ids)
      .includes(all_includes)
      .joins(all_joins)
      .order(order_option)
      .references(:projects)
      .order(sort_criteria_array)
  end

  def work_packages_in_historic_state(timestamp: query.timestamps.first)
    work_packages_in_current_state.at_timestamp(timestamp)
  end
end
```

### (b) Always return the current work packages

As the first approach would always return readonly objects whenever including historic timestamps, and may have also other unintended side effects, I'm leaning towards a more explicit convention:

```ruby
historic_date = 1.year.ago

query = Query.new
query.add_filter 'description', '~', "foo"
query.timestamps = [historic_date]

# This returns the work packages in their current state:
query.results.work_packages

# The historic state needs to be requested explicitly:
query.results.work_packages.at_timestamp(historic_date)
```

How can the historic state be retrieved?

```ruby
# directly
query.results.work_packages.at_timestamp(historic_date)

# using pluck id
WorkPackage.at_timestamp(historic_date).where(id: query.results.work_packages.pluck(:id))

# using a sub query
WorkPackage.at_timestamp(historic_date).where(id: query.results.work_packages)
```

- [x] The sub-query variant does not work, yet.
  - [x]  Fix it ➝ https://github.com/fiedl/openproject/commit/47c45b6a9f8a210eaa59fa7139f31cb69eb016fd
  - [x] Add specific specs ➝ https://github.com/opf/openproject/pull/11678/commits/cadd26e6fe4befade82f2050cc4d38076fb910a6

fiedl committed:
Merge https://github.com/opf/openproject/pull/11243 into https://github.com/opf/openproject/pull/11678

fiedl committed:
`Query::Results`: fixing whitespace [ci-skip]

fiedl committed:
`Query::Results`: always return current, non-historic work-package data

See: https://github.com/opf/openproject/pull/11678#issuecomment-1324244907

fiedl committed:
`Query::Results`: fixing fragile specs

fiedl committed:
`Query::Results`: adding missing joins in the context of historic queries

fiedl commented:
## Checking other unit tests

All specs that are no feature specs are considered unit tests in CI.

```
[2022-11-24 09:11:33] fiedl@data ~/rails/openproject query-timestamps ⚡ 0a9867ce82
▶ docker-compose run backend-test bash
rspec spec/cells
rspec spec/constants
rspec spec/contracts
rspec spec/controllers
rspec spec/decorators
rspec spec/helpers
rspec spec/lib
rspec spec/mailers
rspec spec/migrations
rspec spec/permissions
rspec spec/policies
rspec spec/requests
rspec spec/routing
rspec spec/security
rspec spec/seeders
rspec spec/services
rspec spec/tasks
rspec spec/validator
rspec spec/workers
```

LDAP specs do not work for me locally. All other specs are green, except one:

```
rspec ./spec/workers/application_job_spec.rb:69
# smtp_port is not writable but can be set through env vars or configuration.yml file.
```

This doesn't seem to be feature-related.

There are more specs in `modules/*/spec`. One does fail:

### Distinct Results

```
rspec modules/storages/spec/requests/api/v3/work_packages/work_packages_linked_filter_spec.rb:143
```

The spec expects two work packages to be returned, but they are both returned twice.

I'm trying to reproduce this query:

```ruby
# pry
# modules/storages/spec/requests/api/v3/work_packages/work_packages_linked_filter_spec.rb:143

query = Query.new
query.add_filter 'file_link_origin_id', '=', [origin_id_value]
query.name = "Foo"
query.include_subprojects = false

query.valid?  # => true
query.results.work_packages.count # => 2
query.results.work_packages.uniq.count # => 1

query.results.work_packages.pluck(:id) # => [11238, 11238]

# expected is:
[work_package1, work_package3].map(&:id) # => [11238, 11240]
```

This reproduces the work package being returned twice by the query.

The problem is as follows: The query has a join with two matches as two file links have the same origin id. While in the original implementation the `query.statement` narrows down the search by project conditions etc., the new implementation does that in a sub query. However, the join is still kept to have all table definitions for sorting. This leads, however, to the duplicate entries — one row for each match in the join table.

From a high level, there are at least two ways to resolve this:

- Introduce a `distinct`, which might, however, have other side effects
- Remove the joins not required from the outer query ➝ This is easer than thought because `Query::Results` already provides a separate `sort_criteria_joins` methods. This fixes the issue. ➝ https://github.com/opf/openproject/pull/11678/commits/66907095c6e8b7e354d23e2135cbc28860a7eba8

fiedl committed:
`Query::Results` only join sort-criteria tables in the outer sql query

which fixes an issue with duplicate rows due to surplus joins
https://github.com/opf/openproject/pull/11678#issuecomment-1326171087

fiedl commented:
## Feature Specs

Some feature specs still expose issues with sql queries, which, apparently, are not covered by model specs, yet.

### Eager-loading column aliases

```
rspec './spec/features/work_packages/details/relations/hierarchy_spec.rb[2:1:1:1]'
```

Due to eager loading ([issue](https://github.com/rails/rails/issues/15185), [stackoverflow](https://stackoverflow.com/a/37844262/2066546)), the columns of the `work_package_journals` table are expanded, but, apparently based on the `column_names` of the `WorkPackage` model. But the `work_package_journals` table does not have all columns of the `work_packages` table, which makes the query fail.

```sql
SELECT \"work_package_journals\".\"id\" AS t0_r0, \"work_package_journals\".\"type_id\" AS t0_r1, \"work_package_journals\".\"project_id\" AS t0_r2, \"work_package_journals\".\"subject\" AS t0_r3, \"work_package_journals\".\"description\" AS t0_r4, \"work_package_journals\".\"due_date\" AS t0_r5, \"work_package_journals\".\"category_id\" AS t0_r6, \"work_package_journals\".\"status_id\" AS t0_r7, \"work_package_journals\".\"assigned_to_id\" AS t0_r8, \"work_package_journals\".\"priority_id\" AS t0_r9, \"work_package_journals\".\"version_id\" AS t0_r10, \"work_package_journals\".\"author_id\" AS t0_r11, \"work_package_journals\".\"lock_version\" AS t0_r12, \"work_package_journals\".\"done_ratio\" AS t0_r13, \"work_package_journals\".\"estimated_hours\" AS t0_r14, \"work_package_journals\".\"created_at\" AS t0_r15, \"work_package_journals\".\"updated_at\" AS t0_r16, \"work_package_journals\".\"start_date\" AS t0_r17, \"work_package_journals\".\"responsible_id\" AS t0_r18, \"work_package_journals\".\"budget_id\" AS t0_r19, \"work_package_journals\".\"position\" AS t0_r20, \"work_package_journals\".\"story_points\" AS t0_r21, \"work_package_journals\".\"remaining_hours\" AS t0_r22, \"work_package_journals\".\"derived_estimated_hours\" AS t0_r23, \"work_package_journals\".\"schedule_manually\" AS t0_r24, \"work_package_journals\".\"parent_id\" AS t0_r25, \"work_package_journals\".\"duration\" AS t0_r26, \"work_package_journals\".\"ignore_non_working_days\" AS t0_r27, \"projects\".\"id\" AS t1_r0, \"projects\".\"name\" AS t1_r1, \"projects\".\"description\" AS t1_r2, \"projects\".\"public\" AS t1_r3, \"projects\".\"parent_id\" AS t1_r4, \"projects\".\"created_at\" AS t1_r5, \"projects\".\"updated_at\" AS t1_r6, \"projects\".\"identifier\" AS t1_r7, \"projects\".\"lft\" AS t1_r8, \"projects\".\"rgt\" AS t1_r9, \"projects\".\"active\" AS t1_r10, \"projects\".\"templated\" AS t1_r11 
FROM \"work_package_journals\" 
INNER JOIN \"journals\" ON \"journals\".\"data_type\" = 'Journal::WorkPackageJournal'            
AND \"journals\".\"data_id\" = \"work_package_journals\".\"id\" 
LEFT OUTER JOIN \"projects\" ON \"projects\".\"id\" = \"work_package_journals\".\"project_id\" 
INNER JOIN (SELECT id, created_at FROM \"work_packages\") AS journables            
ON \"journables\".\"id\" = \"journals\".\"journable_id\" 
WHERE \"work_package_journals\".\"project_id\" IN (SELECT \"projects\".\"id\" FROM \"projects\" INNER JOIN \"enabled_modules\" ON \"projects\".\"id\" = \"enabled_modules\".\"project_id\" AND \"enabled_modules\".\"name\" IN ('work_package_tracking') AND \"projects\".\"active\" = TRUE WHERE \"projects\".\"active\" = TRUE) 
AND ((work_package_journals.parent_id IN ('15902'))) 
AND \"journals\".\"id\" IN (SELECT distinct on (journable_type, journable_id) id FROM \"journals\" WHERE \"journals\".\"created_at\" <= '2022-11-26 09:00:38.698132' ORDER BY \"journals\".\"journable_type\" ASC, \"journals\".\"journable_id\" ASC, \"journals\".\"created_at\" DESC)
```

The aliasing is done here: https://github.com/rails/rails/blob/7-0-stable/activerecord/lib/active_record/associations/join_dependency.rb#L168

Using a monkey patch, I'm looking up the call stack where the columns are actually requested.

```ruby
module ActiveRecord
  module Associations
    class JoinDependency

      def aliases
        binding.pry
        
        # p caller
      end

    end
  end
end
```

As workaround, we could deactivate eager loading for these calls:

```ruby
class Journable::HistoricActiveRecordRelation
  def eager_loading?
    false
  end
end
```

### Missing table `work_packages`

In this scenario, the `arel` still has references to `work_packages.id`.

```ruby
           #<Arel::Nodes::Grouping:0x00007f90d7c8ddf8
            @expr=
             #<Arel::Nodes::Or:0x00007f90d7c8de20
              @left=
               "work_packages.id NOT IN (WITH RECURSIVE related (id, from_hierarchy) AS ( SELECT * FROM (VALUES(15956, false)) AS t(id, from_hierarchy) UNION SELECT relations.id, relations.from_hierarchy FROM related JOIN LATERAL ( SELECT CASE WHEN work_package_hierarchies.ancestor_id = related.id THEN work_package_hierarchies.descendant_id ELSE work_package_hierarchies.ancestor_id END id, true from_hierarchy FROM work_package_hierarchies WHERE related.from_hierarchy = false AND (work_package_hierarchies.ancestor_id = related.id OR work_package_hierarchies.descendant_id = related.id) UNION SELECT to_id id, false from_hierarchy FROM relations WHERE (relations.from_id = related.id AND relations.relation_type = 'follows') ) relations ON 1 = 1 ) SELECT id FROM related)",
              @right=
               #<Arel::Nodes::In:0x00007f90d7c8e1b8
                @left=
                 #<struct Arel::Attributes::Attribute
                  relation=
                   #<Arel::Table:0x00007f91089d29c0
                    @klass=
                     WorkPackage(id: integer, type_id: integer, project_id: integer, subject: string, description: text, due_date: date, category_id: integer, status_id: integer, assigned_to_id: integer, priority_id: integer, version_id: integer, author_id: integer, lock_version: integer, done_ratio: integer, estimated_hours: float, created_at: datetime, updated_at: datetime, start_date: date, responsible_id: integer, budget_id: integer, position: integer, story_points: integer, remaining_hours: float, derived_estimated_hours: float, schedule_manually: boolean, parent_id: integer, duration: integer, ignore_non_working_days: boolean),
                    @name="work_packages",
```

The previous implementation did not catch this because the statement was embedded within an `Arel::Nodes::Grouping`. I'm creating also a model spec to cover this case.

### Notification-Center Specs (TimeCop)

There is a failing high-level spec `./spec/features/notifications/notification_center/notification_center_date_alerts_spec.rb`.

As it turns out, this spec changes time using `travel_to`, but does not adjust the journal timestamps. Therefore, the historic work packages are not found because they do not exist at the time travelled to.

fiedl committed:
Merge https://github.com/opf/openproject/pull/11243 into https://github.com/opf/openproject/pull/11678

fiedl committed:
Merge https://github.com/opf/openproject/pull/11243 into https://github.com/opf/openproject/pull/11678

fiedl committed:
Merge https://github.com/opf/openproject/pull/11243 into https://github.com/opf/openproject/pull/11678

fiedl committed:
apply rubocop corrections

fiedl committed:
Merge https://github.com/opf/openproject/pull/11243 into https://github.com/opf/openproject/pull/11678

fiedl committed:
historic relations: disable eager loading

as a workaround for https://github.com/opf/openproject/pull/11678#issuecomment-1328011996

fiedl commented:
## Rubocop

Run rubocop on all files that have changed in this feature branch:

```
[2022-12-07 12:36:32] fiedl@data ~/rails/openproject query-timestamps ⚡ c0f59081d7
▶ docker-compose run backend-test bash
dev@backend-test:~/openproject$ git diff --name-only HEAD $(git merge-base HEAD dev) |xargs rubocop
```

fiedl committed:
apply rubocop corrections

https://github.com/opf/openproject/pull/11678#issuecomment-1340910960

fiedl committed:
Query#timestamps: adding spec with several work packages

https://github.com/opf/openproject/pull/11678

fiedl committed:
Merge https://github.com/opf/openproject/pull/11243 into https://github.com/opf/openproject/pull/11678

fiedl committed:
Query#timestamps: add specs for retrieving historic state of results

https://github.com/opf/openproject/pull/11678#issuecomment-1324244907

fiedl committed:
Query::Timestamps: fixing inconsistent spec

This spec uses `travel_to` to travel to a specific time, but did not adjust the `created_at` column of the journal entries. Therefore, historic queries did not find the work packages because they did not exist at the given time.

The spec did only succeed when executed shortly after midnight, because then the `created_at` column did match by lucky accident.

https://github.com/opf/openproject/pull/11678#issuecomment-1328011996

fiedl committed:
Query::Timestamps: use sub queries rather than collecting ids

as presented in https://github.com/opf/openproject/pull/11243.

Pull request: https://github.com/opf/openproject/pull/11678

fiedl committed:
Query::Timestamps: apply rubocop 1.40.0 corrections

fiedl committed:
Query::Timestamps: implement switch to only use historic queries when historic timestamps are given

https://github.com/opf/openproject/pull/11678

fiedl committed:
Timestamp: adding missing copyright preamble

fiedl commented:
@ulferts

Hi Jens, the second pull request concerning the baseline-comparison feature is ready for review.

1. First pull request: https://github.com/opf/openproject/pull/11243. This provides `WorkPackage.at_timestamp` to query work packages based on the `work_package_journals` table.
2. Second pull request (this pull request): This provides `Query#timestamps=` to configure a query to also filter based on historic data, and modifies the `Query::Results` to search historic data if any historic timestamp is configured. If no historic timestamp is configured, the previous mechanism is used. So, if no `timestamps` are configured, this pull request does not change the behaviour of the application.

Next steps:

- Extend API to accept a `timestamps` parameter
- Extend API to also return historic data of work packages
- Add indices to the `work_package_journals` table

ulferts commented on app/models/journable/historic_active_record_relation.rb:
```diff
@@ -0,0 +1,294 @@
+
+  def eager_loading?
+    false
+  end
```
What is this flag used for?

fiedl replied:
When `eager_loading?` returns `false`, sql queries use `"SELECT * FROM …"`. When `eager_loading?` returns `true`, however, the column names are referenced in sql queries.

However, when switching to the `work_package_journals` table under the hood, not all columns of the `WorkPackage` model are present in the `work_package_journals` table, causing some queries to fail.

Overwriting `eager_loading?` is a quick fix. I've included this as TODO in the feature overview: https://github.com/fiedl/openproject/issues/4

Discussion: https://github.com/opf/openproject/pull/11678#issuecomment-1328011996

ulferts commented on app/models/query/results.rb:
```diff
@@ -39,18 +39,19 @@ def initialize(query)
-      .count
-  rescue ::ActiveRecord::StatementInvalid => e
-    raise ::Query::StatementInvalid.new(e.message)
+    work_packages.count
```
What is the reason for removing the custom error being raised? TBH, I don't know if we really need to have the custom error but this changes the interface and I would like to understand why.

ulferts replied:
The change above will decrease the performance of the count method since e.g.

* more tables will be joined (`joins(all_filter_joins)` vs `joins(all_joins)`)
* ~order is applied~ AR seems to smart enough to remove orders

fiedl replied:
*Phone call 2023-02-02, Ulferts-Fiedlschuster*

- [x] Drop the `work_package_count` method ➝ https://github.com/opf/openproject/pull/11678/commits/c037284a95382cc32dd799340f650eb5d60e6a5e
- [x] Substitute the call in the only occurrence ➝ https://github.com/opf/openproject/pull/11678/commits/c037284a95382cc32dd799340f650eb5d60e6a5e

ulferts commented on app/models/query/results.rb:
```diff
@@ -72,6 +83,31 @@ def order_option
+      .where(query.statement)
+      .includes(all_includes)
+      .joins(all_joins)
+      .references(:projects)
```
Can we please rename this method because I thought it was tasked with getting the relations of a work packages when instead it returns an AR relation on work packages.

fiedl replied:
➝ https://github.com/opf/openproject/pull/11678/commits/8ddea8e5bbf38e14a519af715bc2aaf3e04de514

ulferts commented on app/models/query/results.rb:
```diff
@@ -72,6 +83,31 @@ def order_option
+      .includes(all_includes)
+      .joins(all_joins)
+      .references(:projects)
+  end
```
The three methods

* `work_packages_relation`
* `sorted_work_packages_matching_the_filters_today`
* `sorted_work_packages_matching_the_filters_at_any_of_the_given_timestamps`

share a couple of calls and when understanding the differences I had to disregard those shared calls. I believe some could be moved to the `work_packages_scope` method.

fiedl replied:
➝ https://github.com/opf/openproject/pull/11678/commits/63a55ec24a7a9a24f68d9ad5eebed6b7c9cd2008

ulferts commented on app/models/query/results.rb:
```diff
@@ -72,6 +83,31 @@ def order_option
+    work_package_scope
+      .where(query.statement)
+      .includes(all_includes)
+      .joins(all_joins)
```
`all_joins` includes those [joins only necessary for the sorting](https://github.com/opf/openproject/pull/11678/files#diff-3269fa5bf457650bffd1e9b578998e26f9659d26b3452d823056cbb845a1e8d2L87). Only `all_filter_joins` are necessary I think.

fiedl replied:
Thanks! ➝ https://github.com/opf/openproject/pull/11678/commits/932482578e640b4e4d02148183edf9c0f5530ad7

ulferts commented on app/models/query/results.rb:
```diff
@@ -72,6 +83,31 @@ def order_option
+  def work_packages_relation
+    work_package_scope
+      .where(query.statement)
+      .includes(all_includes)
```
I assume that the includes won't all be applied anyway, since they are not referenced but I also don't think that the line is necessary since apart from the project, only includes for the columns to be displayed are included here which are not necessary to have in the resulting subquery.

fiedl replied:
In the updated implementation, I'm still having `includes(all_includes)` in the `work_pacakge_scope`. Unused includes won't end up in the sql query.

```ruby
WorkPackage.includes(:project).to_sql
=> "SELECT \"work_packages\".* FROM \"work_packages\""

WorkPackage.includes(:project).references(:projects).to_sql
=> "SELECT \"work_packages\".\"id\" AS t0_r0, \"work_packages\".\"type_id\" AS t0_r1, \"work_packages\".\"project_id\" AS t0_r2, \"work_packages\".\"subject\" AS t0_r3, \"work_packages\".\"description\" AS t0_r4, \"work_packages\".\"due_date\" AS t0_r5, \"work_packages\".\"category_id\" AS t0_r6, \"work_packages\".\"status_id\" AS t0_r7, \"work_packages\".\"assigned_to_id\" AS t0_r8, \"work_packages\".\"priority_id\" AS t0_r9, \"work_packages\".\"version_id\" AS t0_r10, \"work_packages\".\"author_id\" AS t0_r11, \"work_packages\".\"lock_version\" AS t0_r12, \"work_packages\".\"done_ratio\" AS t0_r13, \"work_packages\".\"estimated_hours\" AS t0_r14, \"work_packages\".\"created_at\" AS t0_r15, \"work_packages\".\"updated_at\" AS t0_r16, \"work_packages\".\"start_date\" AS t0_r17, \"work_packages\".\"responsible_id\" AS t0_r18, \"work_packages\".\"budget_id\" AS t0_r19, \"work_packages\".\"position\" AS t0_r20, \"work_packages\".\"story_points\" AS t0_r21, \"work_packages\".\"remaining_hours\" AS t0_r22, \"work_packages\".\"derived_estimated_hours\" AS t0_r23, \"work_packages\".\"schedule_manually\" AS t0_r24, \"work_packages\".\"parent_id\" AS t0_r25, \"work_packages\".\"duration\" AS t0_r26, \"work_packages\".\"ignore_non_working_days\" AS t0_r27, \"projects\".\"id\" AS t1_r0, \"projects\".\"name\" AS t1_r1, \"projects\".\"description\" AS t1_r2, \"projects\".\"public\" AS t1_r3, \"projects\".\"parent_id\" AS t1_r4, \"projects\".\"created_at\" AS t1_r5, \"projects\".\"updated_at\" AS t1_r6, \"projects\".\"identifier\" AS t1_r7, \"projects\".\"lft\" AS t1_r8, \"projects\".\"rgt\" AS t1_r9, \"projects\".\"active\" AS t1_r10, \"projects\".\"templated\" AS t1_r11 FROM \"work_packages\" LEFT OUTER JOIN \"projects\" ON \"projects\".\"id\" = \"work_packages\".\"project_id\""
```

Having `includes(all_includes)` within the `work_package_scope` looks clean to me.

However, if you would prefer that, I would suggest that I split up the `include_columns` method so that in `filtered_work_packages` only the filter includes are added, and in `sorted_work_packages` only the sorting includes are added.

ulferts commented on app/models/timestamp.rb:
```diff
@@ -0,0 +1,104 @@
+
+  def self.parse(iso8601_string)
+    if iso8601_string.start_with? "P" # ISO8601 "Period"
+      ActiveSupport::Duration.parse(iso8601_string)
```
If e.g. `Pbar` is the `iso8601_string` value, a `ActiveSupport::Duration::ISO8601Parser::ParsingError` error will be raised. Can we map that to an `ArgumentError` as well to not have to handle too many error types?

fiedl replied:
Actually, an `ActiveSupport::Duration::ISO8601Parser::ParsingError` is a descendant of `ArgumentError`. So catching `ArgumentError` will catch both, `Timestamp.parse("foo")` and `Timestamp.parse("Pfoo")`.

Is that ok or should I catch the `ActiveSupport::Duration::ISO8601Parser::ParsingError` and make it a pure `ArgumentError`?

Added spec: https://github.com/opf/openproject/pull/11678/commits/0ef6fbb11374432769cbc194bc13653be8e6c529

ulferts replied:
No, that is all fine then.

ulferts commented on app/models/query/timestamps.rb:
```diff
@@ -0,0 +1,61 @@
+    # Does this query perform a historic search?
+    #
+    def historic?
+      timestamps != [Timestamp.now]
```
Good approach to have a relative Timestamp be the default value. I was still a bit surprised that it is relative (expecting it to be a concrete value similar to Time.now) but it definitely makes sense to do it like this.

ulferts reviewed (changes requested):
I did not look in depth into the implementation already covered by by #11243 but rather focused on the changes unique to this PR. 

For those, most of the changes are fine as they are quite straight forward. I like the implementation with the Timestamp object. Seems to nicely encapsulate various cases. 

I have no general concerns other than those voiced in #11243 so please find my remarks within the code.

fiedl committed:
Merge https://github.com/opf/openproject/pull/11243 into https://github.com/opf/openproject/pull/11678

fiedl committed:
Query::Results: remove `work_package_count` method

*Phone call 2023-02-02, Ulferts-Fiedlschuster*

- Drop the `work_package_count` method
- Substitute the call in the only occurrence

https://github.com/opf/openproject/pull/11678#discussion_r1064780465

fiedl committed:
Query::Results: move common joins into `work_package_scope`

caught by Jens Ulferts: https://github.com/opf/openproject/pull/11678#discussion_r1064798933

fiedl committed:
Query::Results: improve method naming

caught by Jens Ulferts: https://github.com/opf/openproject/pull/11678#discussion_r1064787719

fiedl committed:
Timestamp: make sure an `ArgumentError` is raised when `Timestamp.parse("Pfoo")`

Caught by Jens Ulferts in https://github.com/opf/openproject/pull/11678#discussion_r1065755175.

fiedl committed:
Query::Results: apply joins only where needed

which removes some redundancy and makes the code more readable.

Caught by Jens Ulferts: https://github.com/opf/openproject/pull/11678#discussion_r1064812318

fiedl committed:
Merge dev into https://github.com/opf/openproject/pull/11678

ulferts commented:
@fiedl with #11243 in, can you please merge `dev` into this PR once more?

fiedl committed:
Merge dev into https://github.com/opf/openproject/pull/11678

ulferts commented on app/models/work_package/pdf_export/work_package_list_to_pdf.rb:
```diff
@@ -84,7 +84,7 @@ def setup_page!
 
   def prepare_batch!
-    total_wp_count = query.results.work_package_count
+    total_wp_count = query.results.work_packages.count
```
Thanks :+1:

ulferts reviewed (approved):
Thanks for the hard work, @fiedl. Looks good to me.

fiedl commented:
Thanks 🎉
